clients/src/test/java/org/apache/kafka/common/metrics/JmxReporterTest.java

```diff
@@ -16,25 +16,123 @@
+            sensor2.add(name, new Total());
+            assertEquals(3, reporter.size());
+            assertEquals(3, reporter.getMBean(name).size());
+            metrics.removeSensor(sensor2.name());
+            assertEquals(3, reporter.size());
+            assertEquals(1, reporter.getMBean(name).size());
+        }
+        assertEquals(0, reporter.size());
+        assertEquals(0, reporter.getMBean(name).size());
+    }
+
+    @Test
+    public void testJmxRegistrationWithMock() throws Exception {
+        final JmxReporter reporter = new JmxReporter();
+        MockKafkaMetric metric;
+        assertEquals(0, reporter.size());
+        metric = MockKafkaMetric.of("pack.bean1.avg", "grp1", new Avg());
+        reporter.metricChange(metric);
+        assertEquals(1, reporter.size());
+        assertEquals(1, reporter.getMBean(metric.metricName()).size());
+        MockKafkaMetric metricAnotherGroup = MockKafkaMetric.of("pack.bean2.total", "grp2", new Total());
+        reporter.metricChange(metricAnotherGroup);
+        assertEquals(2, reporter.size());
+        assertEquals(1, reporter.getMBean(metricAnotherGroup.metricName()).size());
+        List<KafkaMetric> list = new ArrayList<KafkaMetric>();
+        metric = MockKafkaMetric.of("pack.bean1.some", "grp1", new Total());
+        list.add(metric);
+        metric = MockKafkaMetric.of("pack.bean2.some", "grp1", new Total());
+        list.add(metric);
+        reporter.init(list);
+        assertEquals(2, reporter.size());
+        assertEquals(3, reporter.getMBean(metric.metricName()).size());
+        reporter.metricRemoval(metric);
+        assertEquals(2, reporter.size());
+        assertEquals(2, reporter.getMBean(metric.metricName()).size());
+        reporter.metricRemoval(metricAnotherGroup);
+        assertEquals(1, reporter.size());
+        assertEquals(0, reporter.getMBean(metricAnotherGroup.metricName()).size());
+        reporter.close();
+        assertEquals(0, reporter.size());
+        assertEquals(0, reporter.getMBean(metric.metricName()).size());
+    }
+
+    @Test
+    public void testGetMBeanNameWithNull() {
+        final JmxReporter reporter = new JmxReporter();
+        try {
+            reporter.getMBeanName(null);
+        } catch (RuntimeException e) {
+            assertThat(e, CoreMatchers.instanceOf(NullPointerException.class));
+            assertThat(e.getMessage(), CoreMatchers.is("Parameter 'metricName' is mandatory"));
+        }
+    }
+
+    @Test
+    public void testGetMBeanWithNull() {
+        final JmxReporter reporter = new JmxReporter();
+        try {
+            reporter.getMBean(null);
+        } catch (RuntimeException e) {
+            assertThat(e, CoreMatchers.instanceOf(NullPointerException.class));
+            assertThat(e.getMessage(), CoreMatchers.is("Parameter 'metricName' is mandatory"));
+        }
+    }
+
+    private static class MockKafkaMetric extends KafkaMetric {
```

Comment:
The initial motivation was that i could not create an instance of `KafkaMetric` in tests. I've just needed an object to pass it into my `MetricsReporter` implementation.
And basically my mock resolves it - I got a static "fabric" method `of` there: now I'm able to isolate `MetricsReporter` implementation from `Metrics` and test just the case I need.

Also, when I've raised a question on StackOverFlow, @artembilan from Pivotal showed me another approach, which also works:

```java
    Metrics metrics = new Metrics();
    metrics.addMetric("test.name", new Total());
    metrics.metrics().values(); // and here we finally got it :)
```

I've documented the above mentioned approach in javadocs for `KafkaMetric` in scope of this PR.